### PR TITLE
KEYCLOAK-18315 SAML Client - Add parameter to request specific AttributeConsumingServiceIndex

### PR DESCRIPTION
**Pull request by lscorcia**
This PR extracts from KEYCLOAK-17290 - https://github.com/keycloak/keycloak/pull/7817 the required changes to let SAML Clients request a specific attribute set from the remote IdP. As the discussion in https://github.com/keycloak/keycloak/pull/7817 asked for largish required changes to the metadata generation logic, I thought it would make sense to split the two features (new SAML client parameter and metadata generation) into separate PRs.
If/once this PR gets merged I will rebase the other one to only include the changes required for metadata generation, and start working on the suggested changes.

### CODE REVIEW

**Review by mhajas — CHANGES_REQUESTED** [review]
Thank you for the PR @lscorcia! I added few comments in the code.

Just to make sure I understand this correctly. This can be currently used only if an administrator manually adds some `AttributeConsumingService` element into Keycloak's SP metadata? Otherwise, the configuration of the `AttributeConsumingServiceIndex` is basically useless because IDP won't understand it? Is it correct?

File: themes/src/main/resources/theme/base/admin/resources/partials/realm-identity-provider-saml.html [hunk 1]
```diff
@@ -302,6 +302,13 @@
                 </div>
                 <kc-tooltip>{{:: 'identity-provider.allowed-clock-skew.tooltip' | translate}}</kc-tooltip>
             </div>
+            <div class="form-group">
+                <label class="col-md-2 control-label" for="attributeConsumingServiceIndex">{{:: 'identity-provider.saml.attribute-consuming-service-index' | translate}}</label>
+                <div class="col-md-6">
+                    <input class="form-control" string-to-number type="number" min="0" max="2147483" step="1" ng-model="identityProvider.config.attributeConsumingServiceIndex" id="attributeConsumingServiceIndex"/>
```
[the remark this answers is not in the document]
Comment (lscorcia) [reply themes/src/main/resources/theme/base/admin/resources/partials/realm-identity-provider-saml.html:308]: Good catch, I'll update the max value.

File: services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProviderConfig.java [hunk 2]
```diff
@@ -345,6 +346,30 @@ public void setAllowCreated(boolean allowCreate) {
...
+                }
+            } catch (NumberFormatException e) {
+                // ignore it and use null
+            }
+        }
+        return result;
+    }
+
+    public void setAttributeConsumingServiceIndex(Integer attributeConsumingServiceIndex) {
+        if (attributeConsumingServiceIndex < 0) {
```
[the remark this answers is not in the document]
Comment (lscorcia) [reply services/src/main/java/org/keycloak/broker/saml/SAMLIdentityProviderConfig.java:366]: Well, better safe than sorry! I'll add the check.

**Comment by lscorcia** [comment]
Hi @mhajas thanks for the quick review. That's right, if a user wants to specify an attribute set here, currently he should also be adding its info manually to the SP metadata. A SP can also declare more than one requested attribute set, and I don't think Keycloak will support this scenario any time soon - my other patch would add support only for one attribute set, autogenerated from the attribute mappings.

Having this mechanism ready would also help in scenarios where the federation dictates the attribute set indexes to use, e.g. the FICEP gateway to the eIDAS federation (sorry for the buzzword bingo, it's basically the Italian way to join the national SAML ID mechanism with the EU-wide SAML auth). This gateway requires using two specific IDs, 99 and 100, in order to reference two fixed sets of attributes, the minimal one and the full one, even if the SP metadata does not declare them.

**Review by mhajas — APPROVED** [review]
Thanks @lscorcia, LGTM. Also, I think, it makes sense to split the issue into two parts with https://issues.redhat.com/browse/KEYCLOAK-17290 as a follow up.

**Comment by hmlnarik** [comment]
https://master-jenkins.redhat.com/job/universal-test-pipeline-server/570/

**Comment by lscorcia** [comment]
Hello @hmlnarik , I'm sorry but I don't have access to the test pipeline you mentioned. The tests in GitHub pipeline look fine, can you point me to anything more specific?

**Review by hmlnarik — APPROVED** [review]
Thanks for the contribution!

**Comment by hmlnarik** [comment]
> Hello @hmlnarik , I'm sorry but I don't have access to the test pipeline you mentioned. The tests in GitHub pipeline look fine, can you point me to anything more specific?

Hi @lscorcia, no worries, this is an internal pipeline that was running additional tests. No need to check from your side, and they have passed all right.

**Comment by lscorcia** [comment]
Hello @hmlnarik , I just noticed that the documentation PR for this feature has not been merged yet: https://github.com/keycloak/keycloak-documentation/pull/1188 
I have rebased it and it's ready, but I noticed that something is fishy in the asciidoc code above - the "Configuration" and the "Description" columns are reversed after "Principal type" (possibly some merge leftovers?): 

immagine

Could you please take a look at it?

**Pull request merged**